configure-and-exec-dljc.sh

```diff
@@ -0,0 +1,158 @@
+#!/bin/bash
+
+# This script performs WPI via dljc on a given project directory.
+# The inputs are similar to run-dljc.sh, which uses this script internally.
+# The only difference is that run-dljc.sh takes a list of projects, while
+# this script operates on a single project at a time.
+# See the documentation of run-dljc.sh for information on the inputs to this
+# script.
+#
+# Input differences compared to run-dljc.sh:
+# -i and -o are not valid options
+# new option -d: the directory containing the target project
+
+while getopts "c:l:q:s:d:u:t:" opt; do
+  case $opt in
+    c) CHECKERS="$OPTARG"
+       ;;
+    l) CHECKER_LIB="$OPTARG"
+       ;;
+    q) QUALS="$OPTARG"
+       ;;
+    s) STUBS="$OPTARG"
+       ;;
+    d) DIR="$OPTARG"
+       ;;
+    u) USER="$OPTARG"
+       ;;
+    t) TIMEOUT="$OPTARG"
+       ;;        
+    \?) echo "Invalid option -$OPTARG" >&2
+       ;;
+  esac
+done
+
+# check required arguments and environment variables:
+
+# JAVA_HOME must point to a Java 8 JDK for this script to work
+if [ "x${JAVA_HOME}" = "x" ]; then
+    echo "JAVA_HOME must be set to a Java 8 JDK"
+    exit 1
+fi
+
+# testing for JAVA11_HOME, not an unintentional reference to JAVA8_HOME
+# shellcheck disable=SC2153
+if [ "x${JAVA11_HOME}" = "x" ]; then
+    echo "JAVA11_HOME must be set to a Java 11 JDK"
+    exit 1
+fi
+
+JAVA8_HOME="${JAVA_HOME}"
+
+if [ "x${CHECKERFRAMEWORK}" = "x" ]; then
+    echo "CHECKERFRAMEWORK must be set to a locally-built Checker Framework. Please clone and build github.com/typetools/checker-framework"
+    exit 2
+fi
+
+if [ "x${DLJC}" = "x" ]; then
+    echo "DLJC must be set to the dljc executable. Please checkout github.com/kelloggm/do-like-javac and point the DLJC environment variable to its dljc script"
+    exit 3
+fi
+
+if [ ! -d "${DIR}" ]; then
+    echo "configure-and-exec-dljc.sh called on invalid directory: ${DIR}. Please supply a valid directory's absolute path."
+    exit 4
+fi
+
+function configure_and_exec_dljc {
+
+  USABLE="yes"
+  if [ -f build.gradle ]; then
+      chmod +x gradlew
+      BUILD_CMD="./gradlew clean compileJava -g .gradle -Dorg.gradle.java.home=${JAVA_HOME}"
+      CLEAN_CMD="./gradlew clean -g .gradle -Dorg.gradle.java.home=${JAVA_HOME}"
+  elif [ -f pom.xml ]; then
+      # if running on java 8, you must add /jre to the end of this Maven command
+      if [ "${JAVA_HOME}" = "${JAVA8_HOME}" ]; then
+          BUILD_CMD="mvn clean compile -Djava.home=${JAVA_HOME}/jre"
+          CLEAN_CMD="mvn clean -Djava.home=${JAVA_HOME}/jre"
+      else
+          BUILD_CMD="mvn clean compile -Djava.home=${JAVA_HOME}"
+          CLEAN_CMD="mvn clean -Djava.home=${JAVA_HOME}"
+      fi
+  else
+      echo "no build file found for ${REPO_NAME}; not calling DLJC"
+      USABLE="no"
+      return
+  fi
+    
+  DLJC_CMD="${DLJC} -t wpi"
+  if [ ! "x${CHECKERS}" = "x" ]; then
+      TMP="${DLJC_CMD} --checker ${CHECKERS}"
+      DLJC_CMD="${TMP}"
+  fi
+  if [ ! "x${CHECKER_LIB}" = "x" ]; then
```

Comment:
I agree that would have benefits. I'm concerned about distinguishing in a single call to `run-dljc.sh` between arguments that are meant to be passed directly to dljc and arguments that are not (i.e. arguments to `run-dljc.sh` itself). Right now, it's possible to mix the two in any order.

Under your approach, I imagine we would need to add an argument to `run-dljc.sh` that takes as a literal string the arguments to pass to `dljc`. The issue with this is that scripts like `no-literal-securerandom-exact-dljc-cmd.sh` would become very difficult to read, because the classpath arguments to the -q and -l commands would need to all be part of one string. (Unless there's another way to pass all unhandled arguments to another script in bash that I don't know of, in which case we should use that and both of my concerns are moot.)

Reply:
The script could `shift` off all command-line arguments that it recognizes, and pass the rest of the arguments (everything from the first unrecognized command-line argument to the ond) to `dljc`.  This doesn't permit the command-line arguments to be interspersed in arbitrary order, but that does not seem like an important feature to me (feel free to correct me if my assessment is incorrect).